### PR TITLE
Improving dozent progress bar

### PR DESCRIPTION
**Pull request by keelimeguy**

### CODE REVIEW

File: dozent/__init__.py [hunk 1]
```diff
@@ -1,3 +1,6 @@
 from .dozent import Dozent
+from .catch_thread_exceptions import install as _catch_thread_exceptions
+
+_catch_thread_exceptions()
```
Comment (keelimeguy) [inline dozent/__init__.py:4]: https://stackoverflow.com/questions/49663124/cause-python-to-exit-if-any-thread-has-an-exception/49882404#49882404

File: dozent/downloader_tools.py [hunk 2]
```diff
@@ -1,64 +1,88 @@
...
+        prefix = f"[{status}] {dl_size}Mb/{filesize}Mb {f'@{speed}' if speed else ''}"
+        suffix = f"[{int(progress_percentage):3.0f}%{f', {eta} left' if eta else ''}]"
+
+        return progress_percentage, prefix, suffix
 
     @classmethod
-    def download_pysmartdl(cls, link: str, verbose: bool = True):
-        """
-        Downloads file from link using PySmartDL
+    def download_pysmartdl(cls, link: str, download_dir: str, tracker: ProgressTracker = None):
```
Comment (keelimeguy) [inline dozent/downloader_tools.py:52]: I added param for download directory

File: dozent/dozent.py [hunk 3]
```diff
@@ -1,74 +1,115 @@
...
             link = self.queue.get()
             try:
-                DownloaderTools.download_pysmartdl(link)
+                DownloaderTools.download_pysmartdl(link, self.download_dir, tracker=self.tracker)
             finally:
                 self.queue.task_done()
 
 
 class Dozent:
-    def __init__(self, start_date: datetime.datetime, end_date: datetime.datetime):
```
Comment (keelimeguy) [inline dozent/dozent.py:31]: this didn't need to be a class, I better organized the data parsing as well

Comment (keelimeguy) [reply dozent/dozent.py:31]: rather, didn't need to be a class object, static class methods here make more sense

File: dozent/dozent.py [hunk 4]
```diff
@@ -1,74 +1,115 @@
...
+            next(filter(lambda link: (int(link['month']) == end_date.month) and
+                                     (int(link['year']) == end_date.year), data)))
+
+        return [sample_date['link'] for sample_date in data[start_index:end_index]]
+
+    @staticmethod
+    def download_timeframe(start_date: datetime.datetime,
+                           end_date: datetime.datetime,
+                           verbose: bool = True,
+                           download_dir: str = './data'):
```
Comment (keelimeguy) [inline dozent/dozent.py:74]: the default download directory is './data'

File: dozent/dozent.py [hunk 5]
```diff
@@ -1,74 +1,115 @@
...
 if __name__ == "__main__":
+    from catch_thread_exceptions import install as _catch_thread_exceptions
+    _catch_thread_exceptions()
+
     _start_time = time.time()
-    _dozent_object = Dozent(datetime.datetime(2011, 9, 1), datetime.datetime(2016, 10, 1))
-    if _dozent_object.end_date > datetime.datetime(2017, 6, 1):
-        RuntimeError('Not implemented')
-    _dozent_object.download_timeframe()
+    Dozent.download_timeframe(datetime.datetime(2011, 9, 1), datetime.datetime(2016, 10, 1))
```
Comment (keelimeguy) [inline dozent/dozent.py:114]: see how the end-usage becomes much cleaner?

**Review by erb13020 — APPROVED** [review]
Reviewed and looks great! Awesome.

**Pull request merged**